### PR TITLE
(docs) Add Contributing Guidelines

### LINKED ISSUE #2753: Short Contributing primer document
## Problem
I find the size of the hosted docs (readthedocs) overwhelming. I imagine it's simply good documentation, but it's hard for me to skim.

## Solution
I think a succinct primer for contributors would positively influence #1678 (**looking for maintainers & contributors**). Basically, a rudimentary `contributing.md` *on GitHub*. I think it should be on github, because it's much easier to stay on GitHub if you don't need in-depth docs.

I'm envisioning something with the following:
- Links to contributor & user support channels.
- Ways to help (PRs, triage, etc.)
- Important rules for PRs
- Setup-and-run guide

### PR DESCRIPTION
**Pull request by binyamin**
Resolves #2753 
### Changes

Added contributing.md, for the benefit of new contributors.
### Checklist
- [x] Added markup tests, or they don't apply here because... this is just a documentation change
- [x] Updated the changelog at `CHANGES.md` - see above
- [x] Added myself to `AUTHORS.txt`, under Contributors - see above

### CODE REVIEW

File: CONTRIBUTING.md
Comment (joshgoebel) [inline CONTRIBUTING.md:9]: "help welcome" but more specifiically "beginner friendly"... although somewhere we need to disclaim what we mean by "beginner".

File: CONTRIBUTING.md
Comment (joshgoebel) [inline CONTRIBUTING.md:7]: No objection... `Fixing an Issue (via Pull Request)`?

File: CONTRIBUTING.md
Comment (joshgoebel) [inline CONTRIBUTING.md:12]:
```suggestion
Hi, and welcome to Highlight.js. We're a code-highlighting engine etc., but you probably know that. You want to contribute, and this document is the your guide.
```

Correct naming (this applies everywhere).

File: CONTRIBUTING.md
Comment (joshgoebel) [inline CONTRIBUTING.md:14]: Isn't reporting an issue also contributing?

Perhaps mention creating a 3rd party grammar here (with link to the quick start for that)?

File: CONTRIBUTING.md
Comment (joshgoebel) [inline CONTRIBUTING.md:39]: Build and test if probably long enough that's it's ok to have some short summary here (with a link to build and test)... also want to make sure we're not stepping onto the toes of the README, so I think this needs to focus on "I want to setup as a contributor" rather than "how do I use the library on my webpage" (which should be covered well by the readme).

File: CONTRIBUTING.md
Comment (joshgoebel) [inline CONTRIBUTING.md:30]: This should probably be two lists... "before you open a PR" and then then "when" list...

"Open an issue first, so we can explore and discuss it"

This is worth a paragraph.  I can flesh it out if you just get something down.

File: CONTRIBUTING.md
Comment (joshgoebel) [inline CONTRIBUTING.md:11]: Whitespace please! :)

File: CONTRIBUTING.md
Comment (joshgoebel) [inline CONTRIBUTING.md:3]: "Getting Started" vs "Setup and Run" is confusing to me... is this really more of "Welcome"?  Maybe not, but right now the distinction isn't clear. :-)

File: CONTRIBUTING.md
Comment (joshgoebel) [inline CONTRIBUTING.md:27]: This should definitely link to our great JSFiddle with some instructions on how to use it to provide a test case...

https://jsfiddle.net/ajoshguy/2bmdswn6/

**Review by joshgoebel — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md
Comment (binyamin) [reply CONTRIBUTING.md:3]: "Welcome" sounds right.

File: CONTRIBUTING.md
Comment (binyamin) [reply CONTRIBUTING.md:9]: To go down that Rabbit hole... do we mean beginner to _the project_ or to _coding_?

File: CONTRIBUTING.md
Comment (binyamin) [reply CONTRIBUTING.md:7]: That sounds too verbose to me.

Comment (binyamin) [reply CONTRIBUTING.md:7]: `Fixing an Issue (PR)`?

File: CONTRIBUTING.md
Comment (joshgoebel) [reply CONTRIBUTING.md:9]: Ha, perhaps the whole tags are named poorly.  I mean that for someone who:

- knows Javascript
- has a working familiarity with Regex (you won't get too far without this)
- perhaps has some rough familiarity with the idea of parsing text in general
- willingness to read our grammar documentation and look at other grammars

If I think I could possibly fix something in 5-30 minutes I'd consider it potentially beginner friendly.  (ie it's perhaps a simpler rule change that doesn't get into the nuance of our grammars or complex things).  Or there are some issues only involving HTML/CSS/themes/docs... I would tag those beginner friendly.  I guess I meant "to the project"?

I'm open to renaming the tags or coming up with a different classification system if it would be helpful.

I slap most things that don't affect the core parser with "help wanted"... things that I'd like to do myself or where I think someone might be more of a hindrance than help I just leave untagged (and someone could always join those discussions).

File: CONTRIBUTING.md
Comment (binyamin) [reply CONTRIBUTING.md:14]: This https://highlightjs.readthedocs.io/en/latest/language-guide.html? Or the md document?

Comment (joshgoebel) [reply CONTRIBUTING.md:14]: I was thinking of `extra/3RD_PARTY_QUICK_START.md` which was thrown together to answer "I want to make a 3rd party grammar, what do I do?"... Surely it could be improved also. :)

File: CONTRIBUTING.md
Comment (binyamin) [reply CONTRIBUTING.md:28]: revised

File: CONTRIBUTING.md
Comment (binyamin) [reply CONTRIBUTING.md]: done

File: CONTRIBUTING.md
Comment (binyamin) [reply CONTRIBUTING.md]: I'm not familiar enough with the library to write this. Do you mind writing it?

**Comment by binyamin** [comment]
See https://github.com/highlightjs/highlight.js/pull/2760#discussion_r506686019

Also, I'm not comfortable writing the philosophy section having barely contributed anything 😄. Do you mind taking care of that?

Lastly, I can get to work on the "reporting an issue" section. What kinds of issues are useful, and are there any specific rules?

**Comment by joshgoebel** [comment]
> Also, I'm not comfortable writing the philosophy section having barely contributed anything 😄. Do you mind taking care of that?

Yeah, I was planning to.  Give me the weekend.  I'll try and skeleton in some thoughts on the other stuff as well.

**Review by shaonkabir8 — APPROVED** [review]

**Comment by binyamin** [comment]
@joshgoebel friendly ping (regarding"build & test" and "philosophy" sections)

File: CONTRIBUTING.md
Comment (joshgoebel) [inline CONTRIBUTING.md:62]: Is this redundant now that it's in our PR template or is this just good reinforcing?

File: CONTRIBUTING.md
Comment (joshgoebel) [inline CONTRIBUTING.md:67]: Do we need to link to install node or something here?

File: CONTRIBUTING.md
Comment (joshgoebel) [inline CONTRIBUTING.md:103]: We need to find a way to expand this to make it clear there are multiple potential things to work on... someone could submit a new theme without knowing any JS or regex, etc... When I wrote this i was primarily thinking of fixing grammar bugs, etc (our most common type of issue).

File: CONTRIBUTING.md
Comment (joshgoebel) [reply CONTRIBUTING.md]: I just pulled in the TLDR pretty much, thoughts?

**Comment by joshgoebel** [comment]
@allejo @egor-rogov Any high level thoughts here (or i guess tiny concerns also welcome)?  I'm ok if this is a bit bare-bones initially and we flesh it out over time, but it definitely needs to be "useful" before we merge it.

File: CONTRIBUTING.md
Comment (binyamin) [reply CONTRIBUTING.md:62]: I meant it as reinforcement. We could remove it, if you want.

File: CONTRIBUTING.md
Comment (binyamin) [reply CONTRIBUTING.md:67]: It wouldn't hurt. Although, npm comes with Node.js iirc.

File: CONTRIBUTING.md
Comment (binyamin) [inline CONTRIBUTING.md:19]: A bit long, imho. People need to actually _read_ it 😁.

File: CONTRIBUTING.md
Comment (binyamin) [inline CONTRIBUTING.md:27]: This gives the reader a good overview of features. Maybe give it a heading, so we can link to it elsewhere, eg. `#prerequisites`. This would mean the next section needs a heading too, so it's not under `#feature-requests` (faq?)
```suggestion
### Feature Requests
Feature requests are always welcome, but many times the answer is:

> That's a great idea, you should write a plug-in for that! It doesn't belong in the core library and it's probably something you could do with only a small amount of code.

If you'd like to contribute a feature you should always first open an issue to discuss if the feature belongs in core vs a plug-in.  Often this is a great way to get tips for implemention as well or links to prior disussions with context, etc.
```

Comment (binyamin) [reply CONTRIBUTING.md:27]: Actually, `#reporting-an-issue` looks empty. Maye move the feature stuff there.

File: CONTRIBUTING.md
Comment (binyamin) [inline CONTRIBUTING.md:94]: Tiny aesthetics.
```suggestion
For more details see the docs on [Building and Testing](https://highlightjs.readthedocs.io/en/latest/building-testing.html).
```

File: CONTRIBUTING.md
Comment (binyamin) [inline CONTRIBUTING.md:87]: It's already in the "Building & Testing" docs linked below, which should be enough.

File: CONTRIBUTING.md
Comment (binyamin) [reply CONTRIBUTING.md]: I think it's fine.

File: CONTRIBUTING.md
Comment (joshgoebel) [reply CONTRIBUTING.md:67]: Really, has the world just decided npm is dead or is that more of a personal choice :)  Or does node include NPM now out of the box?

File: CONTRIBUTING.md
Comment (joshgoebel) [reply CONTRIBUTING.md:19]: Really?  I didn't find it that along... I can take a second look... did you imagine just a bullet point list or?   What length do you imagine is "perfect" for such a thing?

File: CONTRIBUTING.md
Comment (joshgoebel) [reply CONTRIBUTING.md:27]: I think of feature requests and issues as entirely different things.

File: CONTRIBUTING.md
Comment (joshgoebel) [reply CONTRIBUTING.md:94]: Ok let's not run these tiny one offs by me... please fix grammar, tweak typos, rewrite small sections as necessary... take the initiative, whatever.  I'm trying to focus more on the 1 mile high view. :-)

File: CONTRIBUTING.md
Comment (joshgoebel) [reply CONTRIBUTING.md:27]: I think feature requests makes a great section.  I mean we have a table of contents so as long as this doesn't get too long I'm ok with it having some meat.

File: CONTRIBUTING.md
Comment (joshgoebel) [reply CONTRIBUTING.md:62]: Feels a bit redundant... (the list)... It could just say "when you open a PR we'll have a small checklist waiting for you... it's easy".  No need to repeat.

File: CONTRIBUTING.md
Comment (binyamin) [reply CONTRIBUTING.md:67]: Node includes npm out-of-the-box.

**Comment by allejo** [comment]
> @allejo @egor-rogov Any high level thoughts here (or i guess tiny concerns also welcome)? I'm ok if this is a bit bare-bones initially and we flesh it out over time, but it definitely needs to be "useful" before we merge it.

This is looking good to me! Just a few minor grammar concerns here and there but I'll hold off of that until this PR is ready to be reviewed.

My only thought right now would be to have the "Setup and Run" section be third instead of last. I personally like trying to figure out if I can build/run a project before I get involved with tackling issues/PRs. But other than that, I think the order and sections included here are good.

**Comment by joshgoebel** [comment]
> My only thought right now would be to have the "Setup and Run" section be third instead of last.

Sounds reasonable.

File: CONTRIBUTING.md
Comment (egor-rogov) [inline CONTRIBUTING.md:31]: Found a reference to the google group on that page... Someone monitors it?
And shouldn't we point users to https://github.com/taufik-nurrohman/highlight.ln.js?

File: CONTRIBUTING.md
Comment (egor-rogov) [inline CONTRIBUTING.md:14]: >adding [a 3rd party grammar

There is also this page: https://highlightjs.readthedocs.io/en/latest/language-contribution.html
which is broader (it includes the link to the 3RD_PARTY_QUICK_START.md)

**Review by egor-rogov — COMMENTED** [review]
I've added some thoughts/suggestions, but overall looks good and useful!

File: CONTRIBUTING.md
Comment (joshgoebel) [reply CONTRIBUTING.md:14]: Probably not, but if we aren't disallowing are you suggest we merely discourage by not mentioning it or actively disallow?  Or find some way to get them maintainers on the hook for maintaining themes?

File: CONTRIBUTING.md
Comment (joshgoebel) [reply CONTRIBUTING.md:16]: I guess it doesn't add much.

File: CONTRIBUTING.md
Comment (joshgoebel) [reply CONTRIBUTING.md:31]: We could very briefly, though I think we need another separate place for "Awesome HLJS" stuff in general.

File: CONTRIBUTING.md
Comment (joshgoebel) [reply CONTRIBUTING.md:14]: Yes, there is already talk of merging them long-term in which case this obviously should link to the consolidated document.  This could mention both if it needed to until then - is that what you're suggesting?

Comment (binyamin) [reply CONTRIBUTING.md:14]: How about "working on a theme" instead of "creating a theme", and linking to a list of existing themes as well?

File: CONTRIBUTING.md
Comment (binyamin) [reply CONTRIBUTING.md:19]: multiple sections is fine. I'm mainly worried I usually don't read documents on the web word-for-word, so it should be a easy to scan.

File: CONTRIBUTING.md
Comment (binyamin) [reply CONTRIBUTING.md:16]:
```suggestion
If you need support , [open a GitHub issue](https://github.com/highlightjs/highlight.js/issues/new/choose).
```

File: CONTRIBUTING.md
Comment (binyamin) [reply CONTRIBUTING.md:87]: actually, I'll reference it explicitly.

**Review by sa-orlov — APPROVED** [review]

**Review by walli55 — APPROVED** [review]

**Comment by binyamin** [comment]
@joshgoebel Update: I know I've let this pull request snowball :facepalm:. There are a couple of suggestion threads which have grown stale, and I'm not sure whether to wait for a response or to disregard them. The "reporting an issue" section needs some TLC. Beyond that, can you outline for me what's keeping it from being merged?

**Comment by joshgoebel** [comment]
Let me have another look-see.

File: CONTRIBUTING.md
Comment (joshgoebel) [inline CONTRIBUTING.md:49]: The label changed to "good first issue".

File: CONTRIBUTING.md
Comment (joshgoebel) [inline CONTRIBUTING.md:42]: Lets link to the URL for creating new issues here.

File: CONTRIBUTING.md
Comment (joshgoebel) [inline CONTRIBUTING.md:74]: Would `build_and_test` be better here?

File: CONTRIBUTING.md
Comment (joshgoebel) [inline CONTRIBUTING.md:81]: `build_and_test_browser` ?

File: CONTRIBUTING.md
Comment (joshgoebel) [inline CONTRIBUTING.md:85]:
```suggestion
There's also a tool for visual testing at [tools/developer.html](https://github.com/highlightjs/highlight.js/blob/master/tools/developer.html). It requires a browser build to be present. (ie `npm run build_browser` first)
```

File: CONTRIBUTING.md
Comment (joshgoebel) [inline CONTRIBUTING.md:28]:
```suggestion
- We are more than a keyword highlighter in that we do make _some_ attempt to understand context...
```

**Comment by joshgoebel** [comment]
Can you work on the small things I just flagged and remove all the "draft" mentions... then I'll got over it myself with one last edit pass and we can merge!

File: CONTRIBUTING.md
Comment (binyamin) [reply CONTRIBUTING.md:49]:
```suggestion
If you feel comfortable with our [prerequisites](#prerequisites), you can grab any issue [marked "good first issue"](https://github.com/highlightjs/highlight.js/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22).
```

File: CONTRIBUTING.md
Comment (binyamin) [reply CONTRIBUTING.md:42]:
```suggestion
If you find a bug or think of an improvement, feel free to [open an issue](https://github.com/highlightjs/highlight.js/issues/new/choose).
```

File: CONTRIBUTING.md
Comment (binyamin) [reply CONTRIBUTING.md:74]: I think it feels more natural to run two standard npm commands, than to run one non-standard command. At least, for a beginner.

File: CONTRIBUTING.md
Comment (binyamin) [reply CONTRIBUTING.md:81]: This should be consistent with whatever we do above.

File: CONTRIBUTING.md
Comment (binyamin) [reply CONTRIBUTING.md:15]:
```suggestion
There are many ways to contribute to Highlight.js. Reporting issues is super useful. Feel free to work on an open issue. You can also help by writing [a plugin](https://highlightjs.readthedocs.io/en/latest/plugin-api.html), adding [a 3rd party grammar](https://github.com/highlightjs/highlight.js/blob/master/extra/3RD_PARTY_QUICK_START.md), or developing [a theme](https://highlightjs.readthedocs.io/en/latest/style-guide.html).
```

File: CONTRIBUTING.md
Comment (binyamin) [inline CONTRIBUTING.md:96]:
```suggestion
Are you someone who...
- knows JavaScript?
- has a working familiarity with Regex? (you won't get too far without this)
- perhaps has some rough familiarity with the idea of parsing text in general?
- is willing to read our grammar documentation and look at other grammars?
- wants to help, but doesn't fit any of these requirements?
```

**Review by joshgoebel — APPROVED** [review]

**Pull request merged**
